package NYT.NTest.NProto;

import "yt/systest/proto/table.proto";

message TOperationInputColumns {
    repeated int32 index = 1;
}

message TOperationOutputColumns {
    repeated TDataColumn columns = 2;
}

message TRowMapper {
    oneof operation {
        TSetSeedRowMapper set_seed = 1;
        TIdentityRowMapper identity = 2;
        TGenerateRandomRowMapper generate_random = 3;
        TConcatenateColumnsRowMapper concatenate_columns = 4;
        TDecorateWithDeletedColumnRowMapper decorate_with_deleted_column = 6;
        TRenameColumnRowMapper rename_column = 7;
    }
}

message TMultiMapper {
    oneof operation {
        TFilterMultiMapper filter = 1;
        TSingleMultiMapper single = 2;
        TRepeatMultiMapper repeat = 3;
        TCombineMultiMapper combine = 4;
    }
}

message TReducer {
    oneof operation {
        TSumReducer sum = 1;
        TSumHashReducer sum_hash = 2;
        TConcatenateColumnsReducer concatenate_columns = 3;
    }
}

////////////////////////////////////////////////////////////////////////////////

message TSetSeedRowMapper {
    required int32 input_column_index = 1;
    required int32 this_seed = 2;
}

message TIdentityRowMapper {
    repeated int32 index = 1;
}

message TGenerateRandomRowMapper {
    required TOperationOutputColumns columns = 2;
}

message TConcatenateColumnsRowMapper {
    repeated TRowMapper inner_operations = 1;
};

message TDecorateWithDeletedColumnRowMapper {
    required string stable_name = 1;
}

message TRenameColumnRowMapper {
    required int32 index = 1;
    required string name = 2;
}

////////////////////////////////////////////////////////////////////////////////

message TFilterMultiMapper {
    required int32 input_column_index = 1;
    required bytes value = 2;
}

message TSingleMultiMapper {
    required TRowMapper inner = 1;
}

message TRepeatMultiMapper {
    required int32 count = 1;
    required TRowMapper inner = 2;
}

message TCombineMultiMapper {
    repeated TRowMapper single_operations = 1;
    required TMultiMapper multi_operation = 2;
}

////////////////////////////////////////////////////////////////////////////////

message TSumReducer {
    required int32 input_column_index = 1;
    required TDataColumn output_column = 2;
}

message TSumHashReducer {
    repeated int32 input_column_index = 1;
    required TDataColumn output_column = 2;
}

message TConcatenateColumnsReducer {
    repeated TReducer operations = 1;
};
